EBCDIC Blog Post

NiallBunting opened the pull request:
This contains the new blog post covering my knowledge of EBCDIC. Hopefully will be useful to some people.

Commit:
Draft ebcdic post

NiallBunting committed:
Write more for the EBCDIC page

Commit:
Accidently deleted the post so add some notes.

Hopefully this still exists on my other laptop still.

NiallBunting committed:
Had the draft locally push it up :)

NiallBunting committed:
Update with the latest changes.

We are slowly getting there.

NiallBunting committed:
Rename the blog post to something a better name.

NiallBunting committed:
s/cobybook/copybook/

NiallBunting committed:
Finish the EBCDIC blog post

This finishes the EBCDIC blog post.

NiallBunting commented:
References to add: https://www.tek-tips.com/viewthread.cfm?qid=44991 http://www.csis.ul.ie/cobol/

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -5,7 +5,7 @@ date:   2019-10-23 18:00:00
 ---
 
-This aims to explain what EBCDIC is and how to use it. If you want to read about it's history have a look at the [wiki page][wikipage] (also take a look at the joke section).
+This aims to explain what EBCDIC is and give an overview how to use it. If you want to read about it's history have a look at the [wiki page][wikipage] (also take a look at the joke section).
```
its rather than it's and comma after history

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -31,52 +31,35 @@ This post is mainly going to deal with fixed width EBCDIC records.
 
 At it's heart EBCDIC is just a character encoding, in simple terms that is how we map a number onto character as computers only store numbers i.e. 61:a, 62:b, 63:c.
```
its rather than it's.
Perhaps liken EBCDIC to ASCII at this point: At its heart, both EBCDIC (Extended Binary Coded Decimal Interchange Code) and ASCII (American Standard Code for Information Interchange) are methods of character encoding. In simple terms, they are a way of translating a binary number sequence into letters and numbers: ie 61 = a, 62 = b, 63 = c.

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -31,52 +31,35 @@ This post is mainly going to deal with fixed width EBCDIC records.
+As you can see in the data, there is no schema baked into this record, so if you received this file you would not know how to parse it. 
 
-NUMBERS ARE STORED AS CHARs?
+The schemas are stored in separate files called copybooks, there are two main ways these records can be composed, either fixed or variable width records. The difference between these is how a record is formatted. A fixed width record will always have the exact same length e.g. fixed width of 512 bytes, means each record would be that length even if it was just a name. 
```
EBCDIC fields are a number of characters wide rather than a number of bytes. It's worth using the same terminology throughout as when we get to explaining copybooks the reader may become confused between bytes and characters.
Add a short paragraph that explicitly states how many bytes in an ASCII and an EBCDIC character, what this means with regards to special character encoding etc. From there you can expand on fixed field widths etc

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -31,52 +31,35 @@ This post is mainly going to deal with fixed width EBCDIC records.
 #### EBCDIC Packing
 
-This is a method that EBCDIC uses to store the sign of a given number (if it's positive or negative).
+EBCDIC may store the numbers as strings, however sometimes it uses different characters than 0-9. These values are called packed and is how EBCDIC encodes a positive or negative number, if the schema defines the number to be packed. 
```
This gives a confusing view of EBCDIC numbers and packing.
EBCDIC uses an 8 byte character set, meaning that every row of an EBCDIC file can be read as a string. With a strong focus on minimizing field widths, one trick that this format employs when encoding numbers is to have the sign (whether it's positive or negative) encoded as part of the string changing the least significant digit from a number to a non numeric character. This saves a character by turning the four character "-100" to the three character "10}"

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
````diff
@@ -97,28 +80,40 @@ Let's start off showing a fictional fixed width copybook.
 ```
 
-WHAT DOES PIC MEAN?
+The spaces are important as that is a different section that is reserved for things such as comments.
````
This sentence doesn't really make a lot of sense - consider revising. Indentation is significant, what space is used for comments etc?

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -97,28 +80,40 @@ Let's start off showing a fictional fixed width copybook.
-WHAT DOES PIC MEAN?
+The spaces are important as that is a different section that is reserved for things such as comments.
+
+Next thing are the numbers these define the hierarchy of the copybook. In this example you have the top-level record and then all the fields contained with in it. Other uses could be and Address field that splits the Line 1, Line 2, and Country. Into sub sections. The numbers themselves don't actually matter, just needs to be larger than the parent.
```
Please re-grammar the first sentence

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -97,28 +80,40 @@ Let's start off showing a fictional fixed width copybook.
+
+Next thing are the numbers these define the hierarchy of the copybook. In this example you have the top-level record and then all the fields contained with in it. Other uses could be and Address field that splits the Line 1, Line 2, and Country. Into sub sections. The numbers themselves don't actually matter, just needs to be larger than the parent.
+
+The next line is the name of the field.
```
The next field?

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -97,28 +80,40 @@ Let's start off showing a fictional fixed width copybook.
+
 There are also some special type information:
 * S : The field is packed
 * (n) : for this length.
```
(n) symbolizes multiple instances of the character that precedes it. 9(2) means 2 9s.

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -97,28 +80,40 @@ Let's start off showing a fictional fixed width copybook.
+These fields then are specific modifiers for specific behaviour.
+
 There are also some special type information:
 * S : The field is packed
```
It's easier to think of S as the sign rather than just indicating that the field is packed

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -97,28 +80,40 @@ Let's start off showing a fictional fixed width copybook.
 There are also some special type information:
 * S : The field is packed
 * (n) : for this length.
 * V : A decimal
```
V is a virtual decimal place. It means that the parser will insert a decimal at this point.
The specification also allows for actual decimal places to be included for fields that aren't condensed in this way

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -97,28 +80,40 @@ Let's start off showing a fictional fixed width copybook.
 * (n) : for this length.
 * V : A decimal
 
+That contains most of the info for most of the records, however sometimes there are additional fields such as `BINARY` in this example. There are many different format for more specific use cases.
```
Maybe make a quick list of all the different things that could be here that form part of the EBCDIC reference?

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -97,28 +80,40 @@ Let's start off showing a fictional fixed width copybook.
+
+#### Filler
+
+In fixed width records filler is often used with the record name being `FILLER`. This is for future expansion or spaces in the data where other companies may use and you should just ignore. The Cobrix library will just completely ignore these sections and you can have multiple name `FILLER`.
```
This is the first time mentioning Cobrix, it should probably be mentioned in the introduction about reading in spark

labiraus commented on _posts/2019-12-09-ebcdic-character-format-guide.markdown:
```diff
@@ -144,12 +139,14 @@ Here is our data.
 To convert the following to file use the following command. If you are following this guide to parse the EBCDIC using Spark, you don't need to worry about the spaces as they will be stripped by the library.
```
Your example has "9" as a two character packed field. It should be "0I"

NiallBunting committed:
Fix up the first bunch of issues.

NiallBunting committed:
Finish EBCDIC post

This contains all the changes I'm going to do. It's not perfect
but don't let perfection be the enemy of good.
